### ISSUE TITLE
D3d12 and Vulkan support?

### ISSUE THREAD

**Issue by oscarbg** [ValveSoftware/openvr#63]
any plans to support it? some ETA (even rough like coming later this year)?

**Comment by yaomtc** [comment]
That's up to the developers of the game engines. This is a just a library for interfacing with VR headsets. "OpenVR is an API and runtime that allows access to VR hardware from multiple 
vendors without requiring that applications have specific knowledge of the 
hardware they are targeting."

**Comment by TheWhiteAmbit** [comment]
oh great misusing issue reports for feature requests...

**Comment by oscarbg** [comment]
How is the correct way for feature requests?

2016-03-25 13:59 GMT+01:00 TheWhiteAmbit <notifications@github.com>:

**Comment by JoeLudwig** [comment]
@oscarbg Both are on our radar. No ETA on when we'll add support. To a certain extent it depends on when the major game engines add support for them.

And this is a perfectly reasonable place to ask this question in my book.

**Comment by TheWhiteAmbit** [comment]
Ok, I have another understanding of GitHub-Issues then. But lets make this a discussion forum, ok...

**Comment by stfairy** [comment]
For now when we submit a frame we have to choose between DirectX and OpenGL.  What about Vulkan?

**Comment by Plagman** [comment]
We are working on adding Vulkan support, but it depends on new driver functionality to be able to share Vulkan surfaces with our runtime. We're working with graphics vendors on bringing that up now.

**Comment by Krovius** [comment]
nVidia added a workaround for Vulkan support.

[GL_NV_draw_vulkan_image](http://developer.download.nvidia.com/assets/gameworks/downloads/secure/Vulkan_Beta_Drivers/GL_NV_draw_vulkan_image.txt?autho=1469303206_a45da7b763b6b81d24db4b5a912f2d81&file=GL_NV_draw_vulkan_image.txt)

You could render everything in Vulkan, blit the image to a OpenGL texture object behind a framebuffer, and just give that to IVRCompositor::Submit().

Unfortunately, it only works with nVidia drivers so far.

**Comment by nedelman** [comment]
It's been almost half a year since the original post and there's still apparently no support in OpenVR for D3D12 or Vulkan. At this point I imagine you guys have some kind of beta functionality that you are using internally. Is there some way to get a private release for testing?

About me: I am developing an engine that uses D3D12 and Vulkan but my Vive/VR work is on hold until this is addressed.

**Comment by fhoenig** [comment]
I'm waiting for this as well. You hear "Vulkan Vulkan!" and "VR VR!" everywhere today but try combine the two and you are out of luck. We planned our engine (http://unbound.io) with Vulkan in mind but cannot move forward yet with that.

**Comment by Silverlan** [comment]
Same here, I've been waiting to implement openvr in my engine (running on Vulkan) ever since Vulkan 1.0 was released.
I'm using a work-around with an offscreen OpenGL context right now, which kind of works for low resolutions, but is useless for actual VR development.

**Comment by fhoenig** [comment]
I'd volunteer to write a an example if you just get me this one texture copy operation!!!

**Comment by bartwe** [comment]
I'd also like some more information on future support of dx12 and/or vulkan

**Comment by oscarbg** [comment]
@Plagman well at least now Nvidia should have some support for Vulkan VR in his latest drivers 372.54
"Added a Windows cross-API interop between Vulkan and DirectX 11. This enables Vulkan app compatibility with existing DirectX 11-based HMD runtimes"

**Comment by cabbibo** [comment]
Would just like to chime in and say this interests me as well!

**Comment by aardappel** [comment]
Yes, already working on porting my rendering layer to Vulkan, but this is a blocker.

**Comment by sherief** [comment]
I'd like to add to this and say that I'm extremely interested in D3D12 support. With explicit residency management and the more explicit / async nature of pipeline compilation D3D12 makes it a lot easier to hit [soft] real time targets consistently. That level of UX is desirable for a project I'm working on.

**Comment by Plagman** [comment]
Vulkan has all that and is shipping imminently in SteamVR, any reason you're not targeting that?

**Comment by sherief** [comment]
Literally the first thing I pointed out was "explicit residency management" and Vulkan doesn't have that and then you come along and say, again literally, "Vulkan has all that". I'm not sure how to respond to that.

D3D12 also allows for some presentation behaviors not available with Vulkan under WDDMv2. There doesn't seem to be an equivalent to swap chain GetPresentStatistics() and some Present1() flags under Vulkan.

**Comment by spayne** [comment]
That you elaborated with some info about WDDMv2 helped me understand where you are coming from.

Are there any performance or latency studies that compare Vulkan or D3D12's ability to meet timing requirements?  I'm curious because I'm wondering how big the problem is and if there are tradeoffs or alternatives.

**Comment by sherief** [comment]
There is no Vulkan alternative for the things I mentioned in the second paragraph above, which makes comparing things a bit difficult to say the least.

TL;DR: "Use Vulkan instead" is neither a practical nor possible answer for certain features.

So, D3D12. What's the status there.

**Comment by Plagman** [comment]
Outside the scope of this issue, but if at some point you want to elaborate on exactly what memory management scenarios are not possible to support with Vulkan as opposed to D3D12, it would be good feedback for the API itself.

Since "presentation" in SteamVR is decoupled from the graphics API, why is swap chain functionality important in this context? For the mirror window?

**Comment by sherief** [comment]
The project support VR and non-VR, and the non-VR part isn't achievable without D3D12. Even if you ignore the swap chain parts, DXGI video budget querying, and residency management, an existing D3D12 codebase's only option to support VR being "Do a Vulkan port" is very impractical to say the least. 

What's the best shot at changing the "oh you don't need that" mentality and getting an answer, even if it's a "No, go look elsewhere"?

**Comment by Plagman** [comment]
To be clear, these were just questions originating from legitimate curiosity on my part on why a developer might prefer targeting D3D12 rather than Vulkan, in order to gather feedback and relay it appropriately. It had nothing to do with how SteamVR might or might not support D3D12 and I have no information on that, sorry for derailing.

**Comment by sherief** [comment]
(╯°□°）╯︵ ┻━┻
